gsy_framework/sim_results/electric_blue/time_series.py

```diff
@@ -0,0 +1,148 @@
+from typing import Callable, Dict
+
+from pendulum import DateTime, Duration
+
+from gsy_framework.enums import AvailableMarketTypes
+from gsy_framework.forward_markets.forward_profile import (
+    ForwardTradeProfileGenerator)
+from gsy_framework.sim_results.electric_blue.aggregate_results import (
+    ForwardDeviceStats)
+
+
+def resample_data(
+        timeseries_data: Dict[DateTime, float],
+        resolution: Duration, aggregator_fn: Callable) -> Dict:
+    """Aggregate data using the specified resolution and aggregator function."""
+
+    result = {}
+
+    sorted_time_slots = sorted(timeseries_data.keys())
+    start_time = sorted_time_slots[0]
+    next_time = start_time + resolution
+
+    time_slots_to_aggregate = []
+    for time_slot in sorted_time_slots:
+        if time_slot >= next_time:
+            result[time_slots_to_aggregate[0]] = aggregator_fn(
+                [timeseries_data[t] for t in time_slots_to_aggregate])
+            time_slots_to_aggregate = [time_slot]
+            next_time += resolution
+        elif time_slot < next_time:
+            time_slots_to_aggregate.append(time_slot)
+
+    if time_slots_to_aggregate:
+        # aggregate any remaining time_slots in the buffer
+        result[time_slots_to_aggregate[0]] = aggregator_fn(
+            [timeseries_data[t] for t in time_slots_to_aggregate])
+
+    return result
+
+
+class ForwardDeviceTimeSeries:
+    """Generate time series data using ForwardDeviceStats object."""
+    def __init__(
+            self, device_stats: ForwardDeviceStats, product_type: AvailableMarketTypes):
+
+        self.device_stats = device_stats
+        self.product_type = product_type
+
+    def generate(self, resolution: Duration):
+        """Generate all time series."""
+        return {
+            "matched_buy_orders_kWh": self._generate_matched_buy_orders_kWh(resolution),
+            "matched_sell_orders_kWh": self._generate_matched_sell_orders_kWh(resolution),
+            "open_sell_orders_kWh": self._generate_open_sell_orders_kWh(resolution),
+            "open_buy_orders_kWh": self._generate_open_buy_orders_kWh(resolution),
+            "all_buy_orders_KWh": self._generate_all_buy_orders_kWh(resolution),
+            "all_sell_orders_kWh": self._generate_all_sell_orders(resolution)
+        }
+
+    def _generate_matched_buy_orders_kWh(self, resolution: Duration) -> Dict:
```

Comment:
I think that the implementation of all these methods could be refactored so that the code can be a bit more DRY. If I see correctly, the construction of the ForwardTradeProfileGenerator, the call to the generate_trade_profile method and the resampling can be abstracted into a single method and used by all 6 respective methods.

Reply:
Fixed.